### PR TITLE
add faiss simsearch backend

### PR DESCRIPTION
**Pull request by fantes**
This PR adds support for incremental similarity search on embeddings.

BUILD:
Use `-DUSE_FAISS=ON` with `cmake`

API:
in parameters.output when calling with `build_index: True`:
- `index_type` : a FAISS factory string ie "IVF16,Flat" for simple flat indexing or  "OPQ32_64,IVF256,PQ32" for complex pre and post processing including compression, see https://github.com/facebookresearch/faiss/wiki/Guidelines-to-choose-an-index
- `index_gpu` : put index for indexing and searching on gpu, default : true
- `index_gpuid` : id of gpu to use for indexing, default (no option) is to use all GPUs 
- `train_samples` : number of samples to use for training index, default 100000 see https://github.com/facebookresearch/faiss/wiki/Guidelines-to-choose-an-index
- `ondisk` : (defaults to true) : put inverted lists on disk and not on memory if possible (*IVF* index_types only can do so)
- `nprobe`: (defaults depend on number of clusters noc : max (2,noc/50)) : number of cluster searched for neighbors, used only in IVF* indexes

To create an index incrementally:
- Use `index:True` for indexing a large bulk of data
- Use `build_index:True` to build the index the first time, on disk.
- `search` is available from this point
- Use `index:True` to add any new content, this content is immediately available to search, but not fully persistent
- Use `build_index:True` to commit any newly added content

### CODE REVIEW

**Comment by beniz** [comment]
FYI

```
deepdetect_PR-641/src/simsearch.cc:338:39: error: ‘faiss::gpu’ has not been declared
         faiss::Index* cindex = faiss::gpu::index_gpu_to_cpu(_findex);
```

**Comment by fantes** [comment]
> FYI
> 
> ```c++
> deepdetect_PR-641/src/simsearch.cc:338:39: error: ‘faiss::gpu’ has not been declared
>          faiss::Index* cindex = faiss::gpu::index_gpu_to_cpu(_findex);
> ```

fixed

File: src/simsearch.h [hunk 1]
```diff
@@ -162,7 +184,67 @@ namespace dd
...
+    bool _ondisk = true;
+
+    std::vector<float> _train_samples;
+
+#ifdef USE_GPU_FAISS
+    bool _gpu = false;
+    faiss::Index * _gpu_index;
+    std::vector<faiss::gpu::GpuResources*> _gpu_res;
+    std::vector<int> _gpuids;
+    int _nprobe = -1;
```
Comment (cchadowitz-pf) [inline src/simsearch.h]: Just FYI - I tried building with `-DUSE_CPU_ONLY=ON` and got a compile-time error that `_nprobe` was not a member of `FaissSE`.
```
In file included from /opt/deepdetect/src/mlservice.h:26:0,
                 from /opt/deepdetect/src/services.h:26,
                 from /opt/deepdetect/src/apistrategy.h:26,
                 from /opt/deepdetect/src/deepdetect.h:25,
                 from /opt/deepdetect/src/deepdetect.cc:22:
/opt/deepdetect/src/mlmodel.h: In member function 'void dd::MLModel::create_sim_search(const int&, dd::APIData)':
/opt/deepdetect/src/mlmodel.h:126:20: error: 'class dd::FaissSE' has no member named '_nprobe'
         _se->_tse->_nprobe = ad.get("nprobe").get<int>();
                    ^
In file included from /opt/deepdetect/src/outputconnectorstrategy.h:116:0,
                 from /opt/deepdetect/src/mlservice.h:27,
                 from /opt/deepdetect/src/services.h:26,
                 from /opt/deepdetect/src/apistrategy.h:26,
                 from /opt/deepdetect/src/deepdetect.h:25,
                 from /opt/deepdetect/src/deepdetect.cc:22:
/opt/deepdetect/src/supervisedoutputconnector.h: In member function 'void dd::SupervisedOutput::finalize(const dd::APIData&, dd::APIData&, dd::MLModel*)':
/opt/deepdetect/src/supervisedoutputconnector.h:521:25: error: 'class dd::FaissSE' has no member named '_nprobe'
         mlm->_se->_tse->_nprobe = ad_in.get("nprobe").get<int>();
                         ^
In file included from /opt/deepdetect/src/outputconnectorstrategy.h:117:0,
                 from /opt/deepdetect/src/mlservice.h:27,
                 from /opt/deepdetect/src/services.h:26,
                 from /opt/deepdetect/src/apistrategy.h:26,
                 from /opt/deepdetect/src/deepdetect.h:25,
                 from /opt/deepdetect/src/deepdetect.cc:22:
/opt/deepdetect/src/unsupervisedoutputconnector.h: In member function 'void dd::UnsupervisedOutput::finalize(const dd::APIData&, dd::APIData&, dd::MLModel*)':
/opt/deepdetect/src/unsupervisedoutputconnector.h:222:25: error: 'class dd::FaissSE' has no member named '_nprobe'
         mlm->_se->_tse->_nprobe = ad_in.get("nprobe").get<int>();
```

Comment (fantes) [reply src/simsearch.h]: that should be fixed, thanks a lot !

File: src/simsearch.cc [hunk 2]
```diff
@@ -22,6 +22,13 @@
 #include "simsearch.h"
 #include "utils/fileops.hpp"
 #include "utils/utils.hpp"
+#ifdef USE_FAISS
+#include "faiss/IndexIVF.h"
+#include "faiss/OnDiskInvertedLists.h"
+#include "faiss/IndexPreTransform.h"
+#include "faiss/index_factory.h"
+#include "faiss/gpu/GpuCloner.h"
```
Comment (cchadowitz-pf) [inline src/simsearch.cc]: Sorry, I think this needs a separate `#ifdef USE_GPU_FAISS` as well :)
```
[ 90%] Building CXX object src/CMakeFiles/ddetect.dir/simsearch.cc.o
In file included from /opt/deepdetect/build/faiss/src/faiss/gpu/GpuIndex.h:12:0,
                 from /opt/deepdetect/build/faiss/src/faiss/gpu/GpuCloner.h:15,
                 from /opt/deepdetect/src/simsearch.cc:30:
/opt/deepdetect/build/faiss/src/faiss/gpu/utils/MemorySpace.h:11:18: fatal error: cuda.h: No such file or directory
```

Comment (fantes) [reply src/simsearch.cc]: wow, sorry for that, and thanks a lot again !

Comment (cchadowitz-pf) [reply src/simsearch.cc]: No worries, thanks for the quick fix!

**Comment by beniz** [comment]
>     * `nprobe`: (defaults depend on number of clusters noc : max (2,noc/50)) : number of cluster searched for neighbors, used only in IVF* indexes

Is this parameter is independent of `build_index:True` ? In other words is it set once and for all for an index (or not) ?

**Pull request merged**